sensor/openvas.go

```diff
@@ -0,0 +1,179 @@
+	}
+
+	if sudo {
+		cmdString = append(cmdString, "sudo", "-n")
+	}
+
+	cmdString = append(cmdString, "openvas", "--scan-start", scan)
+
+	head := cmdString[0]
+	args := cmdString[1:]
+
+	cmd := exec.Command(head, args...)
+
+	err := cmd.Start()
+	if err != nil {
+		return fmt.Errorf("unable to start openvas process: %s", err)
+	}
+	go waitForProcessToEnd(cmd.Process, scan)
+	return nil
+}
+
+// StopScan stops a scan with given scan-ID
+func StopScan(scan string, sudo bool) error {
+	err := processes.removeProcess(scan)
+	if err != nil {
+		return err
+	}
+	log.Printf("%s: Stopping scan.\n", scan)
+
+	cmdString := make([]string, 0)
+
+	if sudo {
+		cmdString = append(cmdString, "sudo", "-n")
+	}
+
+	cmdString = append(cmdString, "openvas", "--scan-stop", scan)
+
+	head := cmdString[0]
+	args := cmdString[1:]
+
+	cmd := exec.Command(head, args...)
+
+	err = cmd.Run()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// EndScan must be called when a Openvas Process succesfully finished
+func EndScan(scan string) {
+	processes.removeProcess(scan)
+	log.Printf("%s: Scan successfully finished.\n", scan)
+}
+
+// GetVersion returns the Version of OpenVAS
+func GetVersion() (string, error) {
+	out, err := exec.Command("openvas", "-V").CombinedOutput()
+	if err != nil {
+		return "", err
+	}
+	split := strings.Split(string(out), "\n")
+	return split[0], nil
+}
+
+// GetSettings returns the Settings of OpenVAS as a map
+func GetSettings() (map[string]string, error) {
+	out, err := exec.Command("openvas", "-s").CombinedOutput()
+	if err != nil {
+		return nil, err
+	}
+	settingsList := strings.Split(string(out), "\n")
+	settingsMap := make(map[string]string)
+	for _, setting := range settingsList {
+		settingSplit := strings.Split(setting, "=")
+		if len(settingSplit) != 2 {
+			continue
+		}
+		settingsMap[settingSplit[0]] = settingSplit[1]
+	}
+	return settingsMap, nil
+}
+
+// LoadVTsIntoRedis starts openvas which then loads new VTs into Redis
+func LoadVTsIntoRedis(loadedChan chan struct{}) {
+	log.Printf("Loading VTs into Redis DB...\n")
+
+	err := exec.Command("openvas", "--update-vt-info").Run()
+	if err != nil {
+		log.Printf("OpenVAS Scanner failed to load VTs: %s", err)
+		return
+	}
+	log.Printf("Finished loading VTs into Redis DB.\n")
+	loadedChan <- struct{}{}
+}
+
+// IsSudo checks for sudo permissions
+func IsSudo() bool {
+	cmd := exec.Command("sudo", []string{"-n", "openvas", "-s"}...)
+	err := cmd.Run()
+	if err != nil {
+		log.Printf("Cannot start openvas as sudo: %s", err)
+		return false
+	}
+	return true
+}
+
+// waitForProcessToEnd gets Called as go-routine after OpenVAS Scan Process was
+// started
+func waitForProcessToEnd(p *os.Process, scan string) {
+	processes.addProcess(scan, p)
+	p.Wait()
+	err := processes.removeProcess(scan)
+	if err == nil {
+		log.Printf("%s: Scan process got unexpectedly stopped or killed.\n", scan)
+		// TODO: Interrupt scan
+		return
```

Comment:
Difficult because  its a meant as a goroutine only called by openvas itself. It's responsible to check if an openvas process got terminated without sending a finished message.

Reply:
Ahh, ok, got it.